### PR TITLE
MediaWiki Integration into Plinth (without SSO)

### PR DESCRIPTION
**Pull request by JosephKiranBabu**
Basic installation of MediaWiki is done. The following tasks are pending.

- [x] Auto-configure the mediawiki deployment so that when the user first hits /mediawiki, the Main page is shown.
- [x] Change the wiki.png which appears as default logo to mediawiki.png
- [ ] Single Sign On using Apache REMOTE_USER header. [Auth_remoteuser](https://www.mediawiki.org/wiki/Extension:Auth_remoteuser) (needs Debian packaging?)
- [ ] Checkbox to [disable user registrations](https://www.mediawiki.org/wiki/Extension:LDAP_Authentication/User_Provided_Information#Deactivating_New_User_Signup_.2F_Registration) (should still allow new Plinth users)

### CODE REVIEW

**Comment by SunilMohanAdapa** [comment]
Installing the wiki is something like this:
```
mkdir /var/lib/mediawiki-db/
chown www-data:www-data /var/lib/mediawiki-db
php install.php --confpath=/etc/mediawiki --dbtype=sqlite --dbpath=/var/lib/mediawiki-db/ --scriptpath=/mediawiki --pass test Wiki admin
```

**Comment by JosephKiranBabu** [comment]
Updated the pull request with auto-configuration for mediawiki.
Currently, the mediawiki instance is public, meaning anyone can edit it just like on Wikipedia.
Maybe we can leave it up to the user to decide whether to make it public or private, by giving a checkbox in the configuration page.

**Comment by JosephKiranBabu** [comment]
# Mediawiki Auth mechanism

Anybody can access mediawiki by going to /mediawiki
An anonymous user's page will look like this. Login and edit options are completely missing.

screenshot from 2017-12-01 18-22-46

A Plinth user will go through SSO and hence will get automatically logged in with the same username and also have edit options.

screenshot from 2017-12-01 18-22-59


The trick here is to create another alias for editing the wiki, like /mediawiki-edit which will be the url that Plinth users will get from the UI.
A user without wiki permissions will still see the mediawiki application on the frontpage but will be taken to a read-only version with no edit or login permissions.

**Comment by JosephKiranBabu** [comment]
This story is fully ready to implement. `Auth_remoteuser` extension must be packaged for Debian.

Table of Contents
─────────────────

1 MediaWiki Auth
.. 1.1 Installing extension
.. 1.2 Settings to add in LocalSettings.php
.. 1.3 Change apache configuration to allow only users from Plinth


1 MediaWiki Auth
════════════════════════

1.1 Installing extension
────────────────────────
Download https://www.mediawiki.org/wiki/Special:ExtensionDistributor?extdistname=Auth_remoteuser&extdistversion=REL1_27
```
  $ tar -xzf Auth_remoteuser-REL1_27-380dbbe.tar.gz -C /var/lib/mediawiki/extensions
```
The above steps are for testing purposes only. In the actual implementation, a Debian package for the extension Auth_remoteuser should be used.

1.2 Settings to add in LocalSettings.php
────────────────────────────────────────
```
# Extension to authenticate by REMOTE_USER header
wfLoadExtension( 'Auth_remoteuser' );

# Disable account creation and enable auto-creation only
$wgGroupPermissions['*']['createaccount'] = false;
$wgGroupPermissions['*']['autocreateaccount'] = true;

# Disable anonymous editing
$wgGroupPermissions['*']['edit'] = false;

# Don't even show a login link in the page for anonymous users
function NoLoginLinkOnMainPage( &$personal_urls ){
    unset( $personal_urls['login'] );
    unset( $personal_urls['anonlogin'] );
    return true;
}
$wgHooks['PersonalUrls'][]='NoLoginLinkOnMainPage';
```

1.3 Change apache configuration to allow only users from Plinth
───────────────────────────────────────────────────────────────
```
Alias /mediawiki /var/lib/mediawiki
Alias /mediawiki-edit /var/lib/mediawiki

<Location /mediawiki>
        Options +FollowSymLinks
        AllowOverride All
        <IfVersion >= 2.3>
                Require all granted
        </IfVersion>
        <IfVersion < 2.3>
                order allow,deny
                allow from all
        </IfVersion>
</Location>

<Location /mediawiki-edit>
    Include includes/freedombox-single-sign-on.conf
</Location>

# some directories must be protected
```

**Comment by JosephKiranBabu** [comment]
# Auth using SAML2

screenshot from 2017-12-05 16-48-25
_A button to Login using SAML in Mediawiki_


## Implementation

SAML2 authentication consists of three components

1. the user's browser, 
2. a Service Provider - mediawiki in this case and
3. an Identity Provider - must be implemented in Plinth


## Service Provider
Mediawiki extension [SimpleSamlAuth](https://www.mediawiki.org/wiki/Extension:SimpleSamlAuth) used with the php library [SimpleSamlPHP](https://packages.debian.org/sid/web/simplesamlphp) can serve as the Service Provider.

## Identity Provider
Plinth can act as an Identity Provider using [djangosaml2idp](https://pypi.python.org/pypi/djangosaml2idp) and the [PySaml2](https://tracker.debian.org/pkg/python-pysaml2) library.

The mediawiki extension and django plugin must be packaged for Debian. 
The IdP and SP will require significant setup work. The instructions for setup are in the respective documentation pages.

**Comment by JosephKiranBabu** [comment]
I've moved the SAML2 Auth to another story. 

In the current implementation, MediaWiki gets installed with a random admin password.
There is an entry in the UI to change the administrator password. The user who installed this app will first set an admin password and then login to Mediawiki.
The admin user then creates the remaining users from within MediaWiki. A direct link to create users is provided in the app description.

screenshot from 2017-12-22 16-08-03

File: plinth/modules/mediawiki/views.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Plinth module for configuring Transmission Server
```
Comment (jvalleroy) [inline plinth/modules/mediawiki/views.py]: Transmission -> MediaWiki

File: plinth/modules/mediawiki/__init__.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    _('MediaWiki is the wiki engine that powers Wikipedia and other WikiMedia \
+    projects. A wiki engine is a program for creating a collaboratively \
+    edited website. You can use MediaWiki to host a wiki-like website, take \
+    notes or collaborate with friends on projects.'),
+    _('This MediaWiki instance comes with a randomly generated administrator \
+    password. You can set a new password in the Configuration section and  \
+    login using the "admin" account. You can then create more user accounts \
+    from MediaWiki itself by going to the <a href="/mediawiki/index.php/\
+    Special:CreateAccount">Special:CreateAccount</a> page'),
+    _('Anyone with access this Wiki can read it. Only users that are logged in \
```
Comment (jvalleroy) [inline plinth/modules/mediawiki/__init__.py]: Should it say "access to this Wiki"?

File: plinth/modules/mediawiki/forms.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+from django import forms
+from django.utils.translation import ugettext_lazy as _
+
+from plinth.forms import ServiceForm
+
+
+class MediawikiForm(ServiceForm):  # pylint: disable=W0232
+    """Mediawiki configuration form"""
+    password = forms.CharField(label=_('Administrator Password'), help_text=_(
```
Comment (jvalleroy) [inline plinth/modules/mediawiki/forms.py]: Should the password field be not required? Or should it already be filled in, if a password was set before?

File: plinth/modules/mediawiki/__init__.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+from plinth import service as service_module
+from plinth import action_utils, actions, frontpage
+from plinth.menu import main_menu
+
+from .manifest import clients
+
+version = 1
+
+managed_services = ['mediawiki']
```
Comment (jvalleroy) [inline plinth/modules/mediawiki/__init__.py]: I don't see this service -- should it be 'mediawiki-jobrunner'?

**Comment by SunilMohanAdapa** [comment]
@JosephKiranBabu is on leave and I am trying to update this pull request for merging before tomorrows release.

**Comment by jvalleroy** [comment]
Added in #1182.

**Pull request closed without merging**